File: packages/nativescript-persona/platforms/android/java/io/nstudio/plugins/persona/NSCPersona.kt
package io.nstudio.plugins.persona

import androidx.activity.result.ActivityResultLauncher
import androidx.appcompat.app.AppCompatActivity
import com.google.gson.Gson
import com.withpersona.sdk2.inquiry.*

class NSCPersona private constructor(template: String?) {
  private var template: String? = null
  private var templateVersion: String? = null
  private var inquiry: String? = null
  private var inquiryInstance: Inquiry? = null

  private var accountId: String? = null
  private var referenceId: String? = null
  private var sessionToken: String? = null
  private var environment: Int? = null
  private var fields: Fields? = null

  var callback: Callback? = null
    set(value) {
      if (field != null && value == null) {
        callbacks.remove(field)
      }
      field = value

      if (value != null) {
        callbacks.add(value)
      }
    }

  interface Callback {
    fun onComplete(inquiryId: String, status: String, fields: String)
    fun onCancel(inquiryId: String?, sessionToken: String?)
    fun onError(error: String)
  }

  class FieldItem(var type: String = "unknown", var value: Any? = null)

  init {
    this.template = template
  }

  fun environment(value: Int?): NSCPersona {
    environment = value
    return this
  }

  fun accountId(id: String?): NSCPersona {
    accountId = id
    return this
  }

  fun referenceId(id: String?): NSCPersona {
    referenceId = id
    return this
  }

  fun sessionToken(token: String?): NSCPersona {
    sessionToken = token
    return this
  }

  fun fields(fields: HashMap<String, Any>): NSCPersona {
    val values = Fields.Builder()
    for (field in fields) {
      when (field.value) {
        is Boolean -> {
          values.field(field.key, field.value as Boolean)
        }
        is String -> {
          values.field(field.key, field.value as String)
        }
        is Int -> {
          values.field(field.key, field.value as Int)
        }
      }
    }

    if (fields.isNotEmpty()) {
      this.fields = values.build()
    }
    return this
  }

  fun fields(fields: Fields?): NSCPersona {
    this.fields = fields
    return this
  }

  fun build(): NSCPersona {
    if (inquiry != null) {
      inquiryInstance = Inquiry.fromInquiry(inquiry!!)
        .sessionToken(sessionToken)
        .build()
    } else if (template != null) {
      var builder = Inquiry.fromTemplate(template!!)
        .accountId(accountId)

      if (environment != null) {
        if (environment == 0) {
          builder.environment(Environment.SANDBOX)
        } else if (environment == 1) {
          builder = builder.environment(Environment.PRODUCTION)
        }
      }

      inquiryInstance = builder
        .fields(fields)
        .referenceId(referenceId).build()
    } else if (templateVersion != null) {

      var builder = Inquiry.fromTemplateVersion(templateVersion!!)
        .accountId(accountId)

      if (environment != null) {
        if (environment == 0) {
          builder.environment(Environment.SANDBOX)
        } else if (environment == 1) {
          builder = builder.environment(Environment.PRODUCTION)
        }
      }

      inquiryInstance = builder
        .fields(fields)
        .referenceId(referenceId)
        .build()
    }

    return this
  }

  fun start() {
    inquiryResult?.launch(inquiryInstance)
  }

  @Synchronized
  @Throws(Throwable::class)
  protected fun finalize() {
    if (callback != null) {
      callbacks.remove(callback)
    }
  }


  companion object {

    private var callbacks: ArrayList<Callback> = ArrayList()

    private val gson = Gson()

    private var inquiryResult: ActivityResultLauncher<Inquiry>? = null

    @JvmStatic
    fun init(activity: AppCompatActivity) {
      inquiryResult = activity.registerForActivityResult(
        Inquiry.Contract()
      ) { result ->

        when (result) {
          is InquiryResponse.Complete -> {
            val ret = HashMap<String, FieldItem>()
            for (field in result.fields) {
              when (val value = field.value) {
                is InquiryField.BooleanField -> {
                  ret[field.key] = FieldItem("boolean", value.value)
                }
                is InquiryField.DatetimeField -> {
                  ret[field.key] = FieldItem("date", value.value)
                }
                is InquiryField.DateField -> {
                  ret[field.key] = FieldItem("date", value.value)
                }
                is InquiryField.FloatField -> {
                  ret[field.key] = FieldItem("number", value.value)
                }
                is InquiryField.IntegerField -> {
                  ret[field.key] = FieldItem("number", value.value)
                }
                is InquiryField.StringField -> {
                  ret[field.key] = FieldItem("string", value.value)
                }
                is InquiryField.UnknownField -> {
                  ret[field.key] = FieldItem()
                }
              }
            }

            for (callback in callbacks) {
              callback.onComplete(result.inquiryId, result.status, gson.toJson(ret))
            }
          }
          is InquiryResponse.Cancel -> {
            for (callback in callbacks) {
              callback.onCancel(result.inquiryId, result.sessionToken)
            }
          }
          is InquiryResponse.Error -> {
            for (callback in callbacks) {
              callback.onError(result.debugMessage)
            }
          }
        }
      }

    }


    @JvmStatic
    fun fromTemplate(template: String): NSCPersona {
      return NSCPersona(template)
    }

    @JvmStatic
    fun fromTemplateVersion(template: String): NSCPersona {
      val persona = NSCPersona(null)
      persona.templateVersion = template
      return persona
    }

    @JvmStatic
    fun fromInquiry(inquiry: String): NSCPersona {
      val persona = NSCPersona(null)
      persona.inquiry = inquiry
      return persona
    }
  }
}

File: packages/nativescript-persona/platforms/ios/src/NSCPersona.swift
import Foundation
import UIKit
import Persona2


@objc(NSCPersonaEnvironment)
public enum NSCPersonaEnvironment: Int, RawRepresentable {
    public typealias RawValue = Int
    case Sandbox
    case Production
    
    func toEnv()-> Environment {
        if(self == .Production){
            return .production
        }
        
        return .sandbox
    }
}


@objc(NSCInquiryField)
@objcMembers
public class NSCInquiryField: NSObject {
    private var field: InquiryField
    public init(field: InquiryField) {
        self.field = field
    }
    
    var type: String {
        get {
            switch(field){
            case .string(_):
                return "string"
            case .int(_):
                return "int"
            case .float(_):
                return "float"
            case .bool(_):
                return "boolean"
            case .date(_):
                return "date"
            case .datetime(_):
                return "datetime"
            case .unknown:
                return "unknown"
            }
        }
    }
    
    var value: Any? {
        get {
            switch(field){
            case .string(let ret):
                return ret
            case .int(let ret):
                return ret
            case .float(let ret):
                return ret
            case .bool(let ret):
                return ret
            case .date(let ret):
                return ret
            case .datetime(let ret):
                return ret
            case .unknown:
                return nil
            }
        }
    }
}

@objc(NSCPersona)
@objcMembers
public class NSCPersona: NSObject, InquiryDelegate {
    
    public func inquiryComplete(inquiryId: String, status: String, fields: [String : InquiryField]) {
        
        let ret = fields.mapValues {
            NSCInquiryField(field: $0)
        }
        onCompleteCallback?(inquiryId, status, ret as NSDictionary)
    }
    
    public func inquiryCanceled(inquiryId: String?, sessionToken: String?) {
        onCancelCallback?(inquiry, sessionToken)
    }
    
    public func inquiryError(_ error: Error) {
        onErrorCallback?(error as NSError)
    }
    
    private var inquiryInstance: Inquiry?
    private var template: String?
    private var templateVersion: String?
    private var inquiry: String?
    private var accountId: String?
    private var referenceId: String?
    private var sessionToken: String?
    private var environment: NSCPersonaEnvironment? = nil
    private var theme: InquiryTheme? = nil
    private var fields: [String : InquiryField]? = nil
    
    private var onCompleteCallback: ((String, String, NSDictionary) -> Void)? = nil
    
    private var onCancelCallback: ((String?, String?) -> Void)? = nil
    
    private var onErrorCallback: ((NSError) -> Void)? = nil
    
    
    var configuration: InquiryConfiguration?
    
    private init(template: String?) {
        self.template = template
    }
    
    private init(templateVersion: String?) {
        self.templateVersion = templateVersion
    }
    
    private init(inquiry: String?) {
        self.inquiry = inquiry
    }
    public static func fromTemplate(template: String) -> NSCPersona {
        return NSCPersona(template: template)
    }
    
    public static func fromTemplateVersion(templateVersion: String) -> NSCPersona {
        return NSCPersona(templateVersion: templateVersion)
    }
    
    public static func fromInquiry(inquiry: String) -> NSCPersona {
        return NSCPersona(inquiry: inquiry)
    }
    
    
    public func environment(_ value: NSCPersonaEnvironment) -> Self {
        environment = value
        return self
    }
    
    public func accountId(_ id: String) -> Self {
        accountId = id
        return self
    }
    
    public func referenceId(_ id: String) -> Self {
        referenceId = id
        return self
    }
    
    public func sessionToken(_ token: String) -> Self {
        sessionToken = token
        return self
    }
    
    public func fields(_ fields: NSDictionary) -> Self {
        var values: [String : InquiryField] = [:]
        for field in fields {
            if(field.value is Bool){
                values[field.key as! String] =  InquiryField.bool(field.value as? Bool)
            }else if(field.value is Int){
                values[field.key as! String] =  InquiryField.int(field.value as? Int)
            }else if(field.value is Float){
                values[field.key as! String] =  InquiryField.float(field.value as? Float)
            }else if(field.value is String){
                values[field.key as! String] =  InquiryField.string(field.value as? String)
            }else if(field.value is Date){
                values[field.key as! String] =  InquiryField.date(field.value as? Date)
            }else{
                values[field.key as! String] =  InquiryField.unknown
            }
        }
        
        if(!values.isEmpty){
            self.fields = values
        }
        return self
    }
    
    public func onComplete(_ complete: ((String, String, NSDictionary) -> Void)?) -> Self {
        self.onCompleteCallback = complete
        return self
    }
    
    public func onCancel(_ cancel: ((String?, String?) -> Void)?) -> Self {
        self.onCancelCallback = cancel
        return self
    }
    
    public func onError(_ error: ((NSError) -> Void)?) -> Self {
        self.onErrorCallback = error
        return self
    }
    
    public func build() -> Self {
        if(inquiry != nil){
            self.configuration = InquiryConfiguration(inquiryId: inquiry!, sessionToken: sessionToken, theme: theme)
            return self
        }else if(template != nil){
            if(accountId != nil){
                self.configuration = InquiryConfiguration(templateId: template!, accountId: accountId!, environment: environment?.toEnv(), fields: fields, theme: theme)
            }else if(referenceId != nil){
                self.configuration =  InquiryConfiguration(templateId: template!, referenceId: referenceId!, environment: environment?.toEnv(), fields: fields, theme: theme)
            }else if(environment != nil){
                if(fields != nil){
                    self.configuration = InquiryConfiguration(templateId: template!, environment: environment?.toEnv(), fields: fields)
                }else {
                    self.configuration = InquiryConfiguration(templateId: template!, environment: environment?.toEnv())
                }
                
            }else{
                self.configuration = InquiryConfiguration(templateId: template!)
            }
            return self
        }else if(templateVersion != nil){
            if(accountId != nil){
                self.configuration = InquiryConfiguration(templateVersion: templateVersion!, accountId: accountId!, environment: environment?.toEnv(), fields: fields, theme: theme)
            }else if(referenceId != nil){
                self.configuration =  InquiryConfiguration(templateVersion: templateVersion!, referenceId: referenceId!, environment: environment?.toEnv(), fields: fields, theme: theme)
            }else if(environment != nil) {
                if(fields != nil){
                    self.configuration = InquiryConfiguration(templateVersion: templateVersion!, environment: environment?.toEnv(), fields: fields)
                }else {
                    self.configuration = InquiryConfiguration(templateVersion: templateVersion!, environment: environment?.toEnv())
                }
            }else {
                self.configuration = InquiryConfiguration(templateVersion: templateVersion!)
            }
            return self
        }
        
        return self
    }
    
    public func start(_ controller: UIViewController) {
        self.inquiryInstance = Inquiry(config: configuration!, delegate: self)
        self.inquiryInstance?.start(from: controller)
        
    }
    
    
}
